pyro/infer/mcmc/hmc.py

```diff
@@ -189,7 +189,7 @@ def cleanup(self):
         self._reset()
 
     def sample(self, trace):
-        z = {name: node["value"] for name, node in trace.iter_stochastic_nodes()}
+        z = {name: node["value"].data for name, node in trace.iter_stochastic_nodes()}
```

Comment:
Should this be `.detach()` rather than `.data`? I'm not sure the difference, and I thought we should always use `.detach()` now.

Reply:
Both are actually the same, and return a new instance with requires_grad=True, sharing the same underlying data. Will change to `.detach()` for consistency.